File: src/main/java/kz/smarthealth/patientservice/controller/PatientController.java
package kz.smarthealth.patientservice.controller;

import jakarta.validation.Valid;
import kz.smarthealth.patientservice.aop.Log;
import kz.smarthealth.patientservice.model.dto.PatientDTO;
import kz.smarthealth.patientservice.service.PatientService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.annotation.Secured;
import org.springframework.security.access.prepost.PostAuthorize;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/v1/patients")
public class PatientController {

    private final PatientService patientService;

    /**
     * Creates new patient
     *
     * @param patientDTO patient data
     * @return newly created patient
     */
    @PostMapping
    @Secured({"ROLE_PATIENT", "ROLE_DOCTOR"})
    @Log
    @ResponseStatus(HttpStatus.CREATED)
    public PatientDTO savePatient(@RequestBody @Valid PatientDTO patientDTO) {
        return patientService.savePatient(patientDTO);
    }

    /**
     * Fetches patient by id
     *
     * @param id patient id
     * @return existing patient
     */
    @GetMapping("/{id}")
    @Secured({"ROLE_ORGANIZATION", "ROLE_PATIENT", "ROLE_DOCTOR"})
    @PostAuthorize("returnObject != null and authentication.principal.username == returnObject.userId.toString() " +
            "or hasRole('ROLE_DOCTOR') or hasRole('ROLE_ORGANIZATION')")
    @Log
    public PatientDTO getPatientById(@PathVariable UUID id) {
        return patientService.getPatientById(id);
    }

    /**
     * Fetches list of patients by user id
     *
     * @param userId user id
     * @return list of patients
     */
    @GetMapping("/by-user-id/{userId}")
    @PreAuthorize("hasRole('ROLE_DOCTOR') or hasRole('ROLE_ORGANIZATION') " +
            "or (authenticated and authentication.principal.username == #userId.toString())")
    @Log
    public List<PatientDTO> getPatientsByUserId(@PathVariable UUID userId) {
        return patientService.getPatientsByUserId(userId);
    }

    /**
     * Deletes patient by id
     *
     * @param id patient id
     */
    @DeleteMapping("/{id}")
    @Secured({"ROLE_DOCTOR", "ROLE_PATIENT"})
    @Log
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deletePatientById(@PathVariable UUID id) {
        patientService.deletePatientById(id);
    }
}

File: src/main/kotlin/kz/samat/patientservice/controller/PatientController.kt
package kz.samat.patientservice.controller

import kz.samat.patientservice.model.Patient
import kz.samat.patientservice.service.PatientService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*

/**
 * REST api for patients CRUD operations
 *
 * Created by Samat Abibulla on 2022/12/13
 */
@RestController
@RequestMapping("/api/v1/patients")
class PatientController @Autowired constructor(private val patientService: PatientService) {

    /**
     * Creates set of patients
     *
     * @param patientSet patients to be created
     * @return set of created patients
     */
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun createPatients(@RequestBody patientSet: List<Patient>): List<Patient> =
        patientService.createPatients(patientSet)

    /**
     * Returns patient by id
     *
     * @param id of patient
     * @return patient found by id
     */
    @GetMapping("/{id}")
    fun getPatientById(@PathVariable id: String): Patient = patientService.getPatientById(id)

    /**
     * Returns set of patients by user id
     *
     * @param userId id of user that own patients
     * @return set of patients
     */
    @GetMapping("/by-user/{userId}")
    fun getPatientsByUserId(
        @PathVariable("userId") userId: String
    ): List<Patient> = patientService.getPatientsByUserId(userId)

    /**
     * Deletes patient by id
     *
     * @param id of patient
     */
    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deletePatientById(@PathVariable id: String) = patientService.deletePatientById(id)

    /**
     * Updates patient by id
     *
     * @param id of patient that must be updated
     * @param patient data that must be applied
     * @return updated patient data
     */
    @PutMapping("/{id}")
    fun updatePatientById(@PathVariable id: String, @RequestBody patient: Patient): Patient =
        patientService.updatePatientById(id, patient)
}